pkg/ovs/openflow/ofctrl_builder.go

```diff
@@ -275,6 +275,15 @@ func (b *ofFlowBuilder) MatchARPOp(op uint16) FlowBuilder {
 	return b
 }
 
+// MatchIPDscp adds match condition for matching DSCP field in the IP header. Note, OVS use TOS to present DSCP, and
+// the field name is shown as "nw_tos" with OVS command line, and the value is calculated by shifting the given value
+// left 2 bits.
+func (b *ofFlowBuilder) MatchIPDscp(dscp uint8) FlowBuilder {
+	b.matchers = append(b.matchers, fmt.Sprintf("nw_tos=%d", dscp<<2))
```

Comment:
What is the different between nw_tos and ip_dscp? Only high 6 bits vs low 6 bits and supported version?

Reply:
In IPv4/IPv6 packet header, the field name is TOS, and it is a 1-byte field. This field has two content: DSCP and ECN, DSCP is using the 2-7 bits, while ECN is using the 0-1 bits. 

For OVS, it uses the field "nw_tos" to describe DSCP, and use "nw_ecn" for ECN. With OVS command line and OpenFlow protocol, OVS actually also supports to input with field "ip_dscp", but as a result it always shows the value of "nw_tos"(it is a uint8 value, and the lowest 2 bits is always 0, and the higher 6 bits are filed with DSCP configured value). To keep the string is the same as what OVS shows (with command "ovs-ofctl dump-flows"), I use the "nw_tos" to display. But I still use the "ip_dscp" field in the FlowMod message, so the OpenFlow client could ignore the bits-shift, it only needs to fill the expected DSCP value.

OVS fields description link: http://www.openvswitch.org/support/dist-docs/ovs-fields.7.txt

Reply:
Got it.